### PR TITLE
Pull improvements

### PR DESCRIPTION
**Pull request by euank**
Improve pulling to be more reliable hopefully. See the two associated commit messages.

### CODE REVIEW

File: agent/engine/docker_container_engine.go [hunk 1]
```diff
@@ -44,6 +44,12 @@ const (
 	stopContainerTimeout    = 1 * time.Minute
 	removeContainerTimeout  = 5 * time.Minute
 	inspectContainerTimeout = 10 * time.Second
+
+	// dockerPullBeginTimeout is the timeout from when a 'pull' is called to when
+	// we expect to see output on the pull progress stream. This is to work
+	// around a docker bug which sometimes results in pulls not progressing.
+	dockerPullBeginTimeout = 45 * time.Second
```
Comment (samuelkarp) [inline agent/engine/docker_container_engine.go]: I think I'd be more comfortable with 5 minutes here.

File: agent/engine/docker_container_engine.go [hunk 2]
```diff
@@ -136,7 +142,8 @@ func (dg *DockerGoClient) pullImage(image string) DockerContainerMetadata {
 	// Workaround for devicemapper bug. See:
 	// https://github.com/docker/docker/issues/9718
 	pullLock.Lock()
-	defer pullLock.Unlock()
+	ttime.Sleep(utils.AddJitter(0, dockerPullJitter))
```
Comment (samuelkarp) [inline agent/engine/docker_container_engine.go]: I'm really concerned here that this "workaround" now isn't: we now have a race condition that makes it possible for two pulls with a common layer to occur just starting at different times.

Comment (euank) [inline agent/engine/docker_container_engine.go]: The workaround already relied on no third party pulling at the same time either... We should try and get some numbers on how good or bad jitter vs mutex is here.

File: agent/engine/docker_container_engine.go (hunk 1 above)
Comment (euank) [inline agent/engine/docker_container_engine.go]: Will change

File: agent/engine/docker_container_engine.go [hunk 3]
```diff
@@ -144,6 +149,10 @@ func (dg *DockerGoClient) pullImage(image string) DockerContainerMetadata {
 		Repository:   image,
 		OutputStream: pullWriter,
 	}
+	timeout := ttime.After(dockerPullBeginTimeout)
+	pullBegan := make(chan bool, 1)
+	pullBeganOnce := sync.Once{}
```
Comment (samuelkarp) [inline agent/engine/docker_container_engine.go]: pullDebugObserved?
pullDebugObservedOnce?
Or just add a comment

File: agent/engine/docker_container_engine.go [hunk 4]
```diff
@@ -162,13 +174,35 @@ func (dg *DockerGoClient) pullImage(image string) DockerContainerMetadata {
...
+	}
+	log.Debug("Pull began for image", "image", image)
+
+	select {
+	case err := <-pullFinished:
+		if err != nil {
+			return DockerContainerMetadata{Error: CannotXContainerError{"Pulled", err.Error()}}
+		}
+		return DockerContainerMetadata{}
+	case <-timeout:
```
Comment (samuelkarp) [inline agent/engine/docker_container_engine.go]: Is this intended to be checked again?

**Comment by samuelkarp** [comment]
:+1:

**Pull request merged**